### PR TITLE
command: Update backend hash value only after a successful migration

### LINKED ISSUE #26259: Backend hash is updated too early
### Terraform Version
```
0.12.29
0.13.2
0.14.0-dev (master)
```

### Terraform Configuration Files
Essentialy this is: `command/testdata/backend-change` dir:  https://github.com/hashicorp/terraform/tree/master/command/testdata/backend-change

### Debug Output
Full output with all steps to reproduce: https://gist.github.com/lsierant/ad4169675b0e6ee9697d1b37333f504d

### Expected Behavior
When backend configuration is changed I expect that failed `terraform init` will not change my cached backend config in `.terraform/terraform.tfstate`. Any subsequent (failed) inits should behave the same (informing on changed backend). 

### Actual Behavior
When `terraform init` fails, the hash in `.terraform/terraform.tfstate` is updated with value calculated from updated config, but cached config remains unchanged. Subsequent `terraform init` fails to recognise changed backend config and uses old cached backend without informing the user. 

This could be dangerous as the user is not expecting terraform is using old backend having another one set in config file (see additional context below). 

### Steps to Reproduce
Please see https://gist.github.com/lsierant/ad4169675b0e6ee9697d1b37333f504d for all steps to reproduce with debug output.

Steps: 
1. Have an existing backend (local or remote) configured locally, with existing state. 
2. Change backend configuration in any way. This could be changed remote bucket or path to a local state in config file or by `-backend-config=`
3. Run `terraform init`. Terraform will ask for confirmation whether to copy existing state to the new backend.
4. Do not answer, but interrupt (^C). Alternatively there could be a problem locking the remote state. Effect will be the same - interrupted init.
5. Run `terraform init` again. It thinks backend was not changed (hashes are equal) and initialises cleanly but uses old backend config!

### Additional Context
This behaviour was caught when I had one environment configured and changed remote bucket along with some env config in order to create a completely new environment. I forgot about existing state and was surprised when terraform init informed me about existing state, so I ^C'd instinctively to find out what happened. After that I have run terraform init which finished cleanly without warning. After that running terraform plan resulted in replacing all infra in old backend, despite having new backend bucket in config. 
When preparing this test case I found out that any interruption will cause this behaviour, i.e. network issues or problems with acquiring a lock. 

### References
#13165 - introduced this early hash change
#26260 - I have prepared a failing test case showing this problem as a draft PR

### PR DESCRIPTION
**Pull request by chrisarcand**
Replaces #29856
Closes #26259
Closes #26260

The `Meta.backend_C_r_S_unchanged()` method was sadly pulling too many duties.

It seems to have originally been used as a method to be called when the backend is not changing, with an extra assumption that if the configured backend's hash doesn't match the one in state, surely the hash should just be updated as an option might have been moved to command line flags.

However, this function was used throughout this file as 'the method to load the initialized (but not necessarily configured) backend', regardless of whether or not it is the same (unchanged). This is in addition to `Meta.backendFromState()`, which is used to load the same thing except in the main codepath of 'init -backend=false'.

These changes separate the concerns of backend_C_r_S_unchanged() by

1) Fetching the saved backend (`savedBackend()`)
2) Updating the hash value in the backend cache when appropriate (either by leaving it to the caller to do themselves or by calling `updateSavedBackendHash()`)

This allows migration codepaths to *not* update the hash value until after a migration has successfully taken place.

Although it would make sense to incoporate `Meta.backendFromState()` in to this shared API, I've left that alone as it's not consequential to the specific bug we're after here.

### CODE REVIEW

File: internal/command/meta_backend.go [hunk 1]
```diff
@@ -1074,23 +1091,16 @@ func (m *Meta) backend_C_r_S_changed(c *configs.Backend, cHash int, sMgr *clista
 	return b, diags
 }
 
-// Initiailizing an unchanged saved backend
-func (m *Meta) backend_C_r_S_unchanged(c *configs.Backend, cHash int, sMgr *clistate.LocalState) (backend.Backend, tfdiags.Diagnostics) {
+// Initiailizing an saved backend from the cache file (legacy state file)
```
Comment (alisdair) [inline internal/command/meta_backend.go:1094]:
```suggestion
// Initializing an saved backend from the cache file (legacy state file)
```

**Review by alisdair — APPROVED** [review]
This makes sense to me! And I appreciate the comment changes.

I tested locally against my repro for #26259 too, ✅

File: internal/command/meta_backend.go (hunk 1 above)
Comment (jbardin) [reply internal/command/meta_backend.go:1094]: also "a saved"? ;)

**Review by jbardin — APPROVED** [review]

File: internal/command/meta_backend.go (hunk 1 above)
Comment (chrisarcand) [reply internal/command/meta_backend.go:1094]: Thank you 🙇

Comment (chrisarcand) [reply internal/command/meta_backend.go:1094]: `Initializing an`  😆

**Pull request merged**